File: algorithms/algorithms-cplusplus/leetcode/Question_0023_Merge_k_Sorted_Lists.cc
#include <vector>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
 public:
  ListNode* mergeKLists(vector<ListNode*>& lists) {
    if (lists.empty()) {
      return NULL;
    }
    size_t len = lists.size();
    for (size_t i = 1; i < len; i *= 2) {
      for (size_t j = 0; j < len - i; j += 2 * i) {
        lists[j] = merge2Lists(lists[j], lists[j + i]);
      }
    }
    return lists[0];
  }

  ListNode* merge2Lists(ListNode* l1, ListNode* l2) {
    ListNode dummy(0);
    ListNode* p = &dummy;
    while (l1 && l2) {
      if (l1->val < l2->val) {
        p->next = l1;
        l1 = l1->next;
      } else {
        p->next = l2;
        l2 = l2->next;
      }
      p = p->next;
    }
    p->next = l1 ? l1 : l2;
    return dummy.next;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0023_Merge_k_Sorted_Lists.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0023_Merge_k_Sorted_Lists {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        for (int i = 1; i < lists.length; i *= 2) {
            for (int j = 0; j < lists.length - i; j += 2 * i) {
                lists[j] = mergeTwoLists(lists[j], lists[j + i]);
            }
        }
        return lists[0];
    }

    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                p.next = l1;
                l1 = l1.next;
            } else {
                p.next = l2;
                l2 = l2.next;
            }
            p = p.next;
        }
        p.next = l1 == null ? l2 : l1;
        return dummy.next;
    }
}
